### ISSUE TITLE
Analytics Android 4.3.0 is not supported – results in a crash on start

### ISSUE THREAD

**Issue by ldiqual** [Appboy/appboy-segment-android#2]
appboy-segment-integration 2.1.0 crashes with analytics-android 4.3.0

### build.gradle

```
implementation 'com.appboy:appboy-segment-integration:2.1.0'
implementation 'com.segment.analytics.android:analytics:4.3.0'
```

### Crashlog

On this line: https://github.com/Appboy/appboy-segment-android/blob/master/src/main/java/com/segment/analytics/android/integrations/appboy/AppboyIntegration.java#L206

```
E/AndroidRuntime: FATAL EXCEPTION: main
                  Process: com.takescoop.android.dogfood, PID: 12123
                  java.lang.ClassCastException: com.segment.analytics.ValueMap cannot be cast to com.segment.analytics.Properties
                      at com.segment.analytics.android.integrations.appboy.AppboyIntegration.track(AppboyIntegration.java:206)
                      at com.segment.analytics.IntegrationOperation$10.run(IntegrationOperation.java:186)
                      at com.segment.analytics.Analytics.performRun(Analytics.java:1463)
                      at com.segment.analytics.Analytics$10.run(Analytics.java:824)
                      at android.os.Handler.handleCallback(Handler.java:739)
                      at android.os.Handler.dispatchMessage(Handler.java:95)
                      at android.os.Looper.loop(Looper.java:145)
                      at android.app.ActivityThread.main(ActivityThread.java:5835)
                      at java.lang.reflect.Method.invoke(Native Method)
                      at java.lang.reflect.Method.invoke(Method.java:372)
                      at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:1399)
                      at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1194)
```

**Comment by chshapiro** [comment]
Hey @ldiqual ,

Unfortunately we were not able to reproduce this crash. Would you be able to provide us with your project or a similar project that reproduces the issue? You can attach it here or email support@braze.com.

Thanks,
Chava

**Comment by ldiqual** [comment]
@chshapiro Thanks a lot for following up on this issue! We're now experiencing this crash with Segment 4.2.6 though it's unclear how to reproduce it consistently.

After digging in the stack trace, we see [this](https://github.com/Appboy/appboy-segment-android/blob/master/src/main/java/com/segment/analytics/android/integrations/appboy/AppboyIntegration.java#L206):

```java
// com.segment.analytics.android.integrations.appboy.AppboyIntegration.track(AppboyIntegration.java:206)
Properties campaignProps = (Properties) properties.get("campaign");
```

This happens when the `Install Attributed` event gets tracked, which we've been unable to reproduce.

So appboy is expecting the "campaign" key to have a `Properties` type. However when you look at this [segment code](https://github.com/segmentio/analytics-android/blob/a6d77636bd19981e749646bf22f26b8f2646ee5c/analytics/src/main/java/com/segment/analytics/AnalyticsContext.java#L223-L225) it actually returns a `ValueMap` that can be converted to a `Campaign` class: https://github.com/segmentio/analytics-android/blob/a6d77636bd19981e749646bf22f26b8f2646ee5c/analytics/src/main/java/com/segment/analytics/AnalyticsContext.java#L329-L400

Most likely what the appboy integration should do is use `properties.campaign()` which will safely use `CAMPAIGN_KEY` and return a `Campaign` class, instead of trying to deserialize manually.

I would love your input on how to reproduce the `Install Attributed` event on our side. In the meantime we'll revert to an appboy integration version earlier than [2.0.1](https://github.com/Appboy/appboy-segment-android/commit/a13d5cb3b29e88cb522516820033600b114b3268) since it is what is causing the crash.

Let me know

**Comment by ldiqual** [comment]
Another important question is which version we should revert `com.appboy:appboy-segment-integration` to. Some of our dependencies like the mixpanel integration rely on Segment 4.2.6+, can we use the 1.1.0 integration with 4.2.6 or will it break? 2.0.1 and 2.0.0 will most likely crash as well for other reasons.

**Comment by chshapiro** [comment]
Hey @ldiqual ,

How are you tracking`Install Attributed` events in your app now? We recommend following Segment's documentation https://segment.com/docs/spec/mobile/#install-attributed where the "campaign" key does have a `Properties` type.

Please let us know if this fixes your crash.

Thanks,
Chava

**Comment by ldiqual** [comment]
@chshapiro Thanks a lot for following up on this! We do not track the `Install Attributed` event manually, I believe segment propagates it automatically:

 >When Segment or an integrated partner can discern the source of an install, we’ll collect an Install Attributed event.

Sorry to insist but the segment code shows that the `campaign` key contains a `ValueMap`, not a `Properties`: https://github.com/segmentio/analytics-android/blob/a6d77636bd19981e749646bf22f26b8f2646ee5c/analytics/src/main/java/com/segment/analytics/AnalyticsContext.java#L223-L225

While we resolve this issue, can you follow up on which version we can safely downgrade the appboy integration to?

**Comment by chshapiro** [comment]
Hey @ldiqual,

Thanks for the quick reply.

I believe the method where Segment tracks install attribution automatically is https://github.com/segmentio/analytics-android/blob/master/analytics/src/main/java/com/segment/analytics/Analytics.java#L344.

You can also see in the Segment documentation that the `campaign` key should point to a `Properties` instance: https://segment.com/docs/spec/mobile/#install-attributed

The campaign object linked here https://github.com/segmentio/analytics-android/blob/master/analytics/src/main/java/com/segment/analytics/AnalyticsContext.java#L329-L400 is a wrapper for Google Analytics campaigns, not install attribution campaigns so shouldn't be passed down with an install attribution track call.

Are you integrating any other Segment partners other than Braze? It's possible that they could be implementing this incorrectly.

Let us know your thoughts or if we are still missing info on our end, hope to get to the bottom of this soon.

Thanks,
Chava

**Comment by ldiqual** [comment]
@chshapiro Really good push here, I didn't think about other integrations potentially messing things up.

We do use the AppsFlyer segment integration and I believe I finally found the origin of the `ValueMap` being passed instead of `Properties` for the `campaign` key: https://github.com/AppsFlyerSDK/AppsFlyer-Segment-Integration/blob/3f2f5f93a31efd51c9e5447a3b280f2ef59df3d3/app/src/main/java/com/segment/analytics/android/integrations/appsflyer/AppsflyerIntegration.java#L199-L202

It seems that the segment docs have changed recently to show an example with a Props instead of a `ValueMap`. However I believe the appboy integration should be a bit more defensive when doing typed deserialization especially since the docs don't specify that integrations should enforce `Props` which is just a subclass of `ValueMap`. The `track` interface accepts a `ValueMap `and I wouldn't rely on examples to infer a subclass of it for a specific key.

What I would propose is to check that the type of the `campaign` value is a `Properties`, then proceed with a cast. Alternatively you could even just cast to a `ValueMap` since it's a compatible type. Would any of those solutions work?

In the meantime I'll open an issue with the AppsFlyer team and see what they take is on all this – maybe they're not yet Segment 4.2+ compatible.

**Comment by chshapiro** [comment]
Hey @ldiqual ,

We're going to release a new version in the next couple of days that will protect against this crash.

We also suggest you open an issue with AppsFlyer and see if they can start following the format that Segment documents.

Thanks,
Chava

**Comment by chshapiro** [comment]
Hey @ldiqual,

We released version 2.1.1 of our SDK, which protects against this crash.

Please let us know if you have any further issues.

Thanks,
Chava

**Comment by Bucimis** [comment]
Hi @ldiqual,

Closing now. Let us know if this issue still exists for you.

Thanks,
Braze SDK Engineering
